proof/crefine/RISCV64/Recycle_C.thy

```diff
@@ -805,16 +805,123 @@ lemma ntfn_q_refs'_no_NTFNBound[simp]:
   "(x, NTFNBound) \<notin> ntfn_q_refs_of' ntfn"
   by (auto simp: ntfn_q_refs_of'_def split: ntfn.splits)
 
+lemma restart_thread_if_no_fault_ccorres:
+  "ccorres dc xfdc
+     (valid_objs' and no_0_obj' and pspace_aligned' and pspace_distinct' and pspace_bounded'
+      and (\<lambda>s. ksCurDomain s \<le> maxDomain) and (\<lambda>s. weak_sch_act_wf (ksSchedulerAction s) s)
+      and tcb_at' t)
+     \<lbrace>\<acute>thread = tcb_ptr_to_ctcb_ptr t\<rbrace> hs
+     (restartThreadIfNoFault t)
+     (Call restart_thread_if_no_fault_'proc)"
+  apply (cinit lift: thread_')
+   apply (rule ccorres_pre_threadGet, rename_tac fault_opt)
+   apply (rule ccorres_move_c_guard_tcb)
+   apply (rule_tac val="case_option (scast seL4_Fault_NullFault) fault_to_fault_tag fault_opt"
+               and xf'=ret__unsigned_longlong_'
+               and R="\<lambda>s. \<exists>tcb. ko_at' tcb t s \<and> tcbFault tcb = fault_opt"
+                in ccorres_symb_exec_r_known_rv_UNIV[where R'=UNIV])
+       apply (vcg, clarsimp)
+       apply (erule (1) cmap_relation_ko_atE[OF cmap_relation_tcb])
+       subgoal
+         by (fastforce simp: ctcb_relation_def typ_heap_simps cfault_rel_def seL4_Fault_lift_def
+                             Let_def
+                      split: if_split_asm option.split)
+     apply ceqv
+    apply (rule ccorres_cond[where R=\<top>])
+      apply (clarsimp split: option.splits fault.splits)
+      apply (rename_tac fault s s', case_tac fault; clarsimp simp: seL4_Faults)
+      apply (rename_tac arch_fault, case_tac arch_fault, clarsimp simp: seL4_Arch_Faults)
+     apply (rule ccorres_rhs_assoc)+
+     apply (ctac add: setThreadState_ccorres)
+       apply (clarsimp simp: ifCondRefillUnblockCheck_def)
+       apply (rule ccorres_pre_threadGet)
+       apply (rule ccorres_move_c_guard_tcb)
+       apply (clarsimp simp: when_def)
+       apply (simp add: if_to_top_of_bind)
+       apply (rule ccorres_if_lhs)
+        apply (clarsimp simp: bind_assoc, rename_tac scPtr)
+        apply (wpfix add: option.sel)
```

Comment:
this looks interesting, what does the `add: option.sel` do here? I might want to pick up this technique

Reply:
I'm sure Gerwin or Corey would be able to give you a better explanation, but this has the effect of moving from a schematic which has `Some scPtr` as one of its arguments, to additionally having `scPtr` as one of its arguments, so that unification can go ahead in the following line. I think providing `wpfix` with other `.sel` rules will do similar things for other datatypes, but I've only used a handful of times myself

Reply:
Yes that is right. I didn't actually know that you could give these to `wpfix` directly as parameters :-) I thought you had to declare them globally. For `option` I thought we had done that already, but it might be in a theory further down or something like that. We should probably add an issue to set up `wpfix` more properly for all of the basic datatypes we might need it for (option, sum, etc).

But separately from this PR.

Reply:
So if we both had no idea about it and assumed that it had proper setup, does that mean that this would work without the `add: option.sel`? Do you happen to know whether that works Michael? If not, we definitely should add it to the default setup of wpfix!

Reply:
Oh it can be removed from the `wpfix`! That's good. I'll go with that